### PR TITLE
SB-887: remote  RSS index download for Nuget remote repository [in progress]

### PR DESCRIPTION
**Pull request by sbespalov**

### CODE REVIEW

File: strongbox-storage/strongbox-storage-core/src/main/java/org/carlspring/strongbox/repository/RepositoryFeatures.java [hunk 1]
```diff
@@ -11,5 +13,9 @@
     // Indexing implementation
 
     // List of supported checksums
-
+    
+    public void downloadRemoteIndex(String storageId,
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-core/src/main/java/org/carlspring/strongbox/repository/RepositoryFeatures.java]: This should not be a method in the interface. In NuGet, this similar to a feed. In NPM and P2 there will be different things specific to those providers.

File: strongbox-commons/src/main/java/org/carlspring/strongbox/artifact/coordinates/AbstractArtifactCoordinates.java [hunk 2]
```diff
@@ -23,13 +23,19 @@ public AbstractArtifactCoordinates()
     public AbstractArtifactCoordinates(Map<String, String> coordinates)
     {
         this.coordinates = coordinates;
+        this.uuid = toPath();
```
Comment (carlspring) [inline strongbox-commons/src/main/java/org/carlspring/strongbox/artifact/coordinates/AbstractArtifactCoordinates.java]: So... what's the purpose of this...?

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-nuget-layout-provider/src/main/java/org/carlspring/strongbox/repository/NugetRepositoryFeatures.java [hunk 3]
```diff
@@ -1,10 +1,112 @@
...
+            int packageCount;
+            try
+            {
+                packageCount = nugetClient.getPackageCount(false);
+            }
+            catch (IOException | URISyntaxException e)
+            {
+                throw new ArtifactTransportException(e);
+            }
+            for (int i = 0; i < packageCount / 100 + packageCount % 100; i++)
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-nuget-layout-provider/src/main/java/org/carlspring/strongbox/repository/NugetRepositoryFeatures.java]: Could you please explain (for my better understanding) what this loop is about?

File: strongbox-storage/strongbox-storage-core/src/main/java/org/carlspring/strongbox/repository/RepositoryFeatures.java (hunk 1 above)
Comment (sbespalov) [reply strongbox-storage/strongbox-storage-core/src/main/java/org/carlspring/strongbox/repository/RepositoryFeatures.java]: I think that it should. 
> In NuGet, this similar to a feed. In NPM and P2 there will be different things specific to those providers.

all of then is indexes: 
- lucene for Maven
- OrientDB for Nuget

The interface is abstraction which is downloading "indexes" from some kind of "Remote source" and no matter what the indexes are and from what source.

File: strongbox-commons/src/main/java/org/carlspring/strongbox/artifact/coordinates/AbstractArtifactCoordinates.java (hunk 2 above)
Comment (sbespalov) [reply strongbox-commons/src/main/java/org/carlspring/strongbox/artifact/coordinates/AbstractArtifactCoordinates.java]: the purpose that `Path` is unique and can be used as `uuid`

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-nuget-layout-provider/src/main/java/org/carlspring/strongbox/repository/NugetRepositoryFeatures.java (hunk 3 above)
Comment (sbespalov) [reply strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-nuget-layout-provider/src/main/java/org/carlspring/strongbox/repository/NugetRepositoryFeatures.java]: it's the iteration for Feed pages

File: strongbox-storage/strongbox-storage-core/src/main/java/org/carlspring/strongbox/repository/RepositoryFeatures.java (hunk 1 above)
Comment (carlspring) [reply strongbox-storage/strongbox-storage-core/src/main/java/org/carlspring/strongbox/repository/RepositoryFeatures.java]: Yes, these are indexes in the case of Lucene and OrientDB. However, for NuGet, they're feeds. I'm not sure what they're called in NPM-speak...

This is why I have implemented the concept of repository features.

We should not assume they're interchangeable terms and that each layout provider will have an index. In this particular case, we're not talking about OrientDB, or the Maven Indexer. These are, in effect different concepts.

Comment (sbespalov) [reply strongbox-storage/strongbox-storage-core/src/main/java/org/carlspring/strongbox/repository/RepositoryFeatures.java]: In this case, I'm not sure that I fully understand for now this Repository Features concept, but I will remove method from interface as you wished.

Comment (sbespalov) [reply strongbox-storage/strongbox-storage-core/src/main/java/org/carlspring/strongbox/repository/RepositoryFeatures.java]: Should I leave the method in `NugetRepositoryFeatures` class or move it somewhere else?

File: strongbox-commons/src/main/java/org/carlspring/strongbox/artifact/coordinates/AbstractArtifactCoordinates.java (hunk 2 above)
Comment (carlspring) [reply strongbox-commons/src/main/java/org/carlspring/strongbox/artifact/coordinates/AbstractArtifactCoordinates.java]: The path is unique to the repository. I'm not sure, if this is relevant... (Just thought of mentioning it, just in case).

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-nuget-layout-provider/src/test/java/org/carlspring/strongbox/storage/indexing/NugetRemoteRepositoryTest.java [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+ * @author Sergey Bespalov
+ *
+ */
+@ContextConfiguration(classes = NugetLayoutProviderTestConfig.class)
+@RunWith(SpringJUnit4ClassRunner.class)
+public class NugetRemoteRepositoryTest
+        extends TestCaseWithRepository
+{
+
+    private static final String REPOSITORY_RELEASES = "m2pr-releases";
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-nuget-layout-provider/src/test/java/org/carlspring/strongbox/storage/indexing/NugetRemoteRepositoryTest.java]: The convention that I've used is that each test should have their own repository. The repository's name should be an abbreviation of the class name's initials. In this case -- `nrrt`.

We need to keep making sure that each of the tests is omnipotent and that none of the test depend on each other.

The same applies for the variable below.

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-nuget-layout-provider/src/main/java/org/carlspring/strongbox/repository/NugetRepositoryFeatures.java [hunk 5]
```diff
@@ -1,10 +1,112 @@
...
         implements RepositoryFeatures
 {
 
+    @Inject
+    private ConfigurationManager configurationManager;
+
+    @Inject
+    private ArtifactEntryService artifactEntryService;
+
+    public void downloadRemoteIndex(String storageId,
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-nuget-layout-provider/src/main/java/org/carlspring/strongbox/repository/NugetRepositoryFeatures.java]: In this case, I assume a better name for this method, would probably be `downloadRemoteFeed`...? This is what I was referring to in my earlier comment regarding the method that you defined in the interface -- in one implementation, the "features" may be called `X`, `Y`, `Z`, while in another -- there could be `Y` and `A`. The `ReposiotryFeatures` interface defines no methods for this exact reason. However, there is a mechanism in place that uses these "features".

File: strongbox-commons/src/main/java/org/carlspring/strongbox/artifact/coordinates/AbstractArtifactCoordinates.java (hunk 2 above)
Comment (sbespalov) [reply strongbox-commons/src/main/java/org/carlspring/strongbox/artifact/coordinates/AbstractArtifactCoordinates.java]: yes, you are right.
seems that i missed that

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-nuget-layout-provider/src/main/java/org/carlspring/strongbox/repository/NugetRepositoryFeatures.java (hunk 5 above)
Comment (sbespalov) [reply strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-nuget-layout-provider/src/main/java/org/carlspring/strongbox/repository/NugetRepositoryFeatures.java]: ok, I will rename it
thanks

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/domain/ArtifactEntry.java [hunk 6]
```diff
@@ -18,7 +21,10 @@
     private String repositoryId;
 
     // if you have to rename this field please update ArtifactEntryServiceImpl.findByCoordinates() implementation
-    private ArtifactCoordinates artifactCoordinates;
+    @OneToOne(orphanRemoval = true)
+    private AbstractArtifactCoordinates artifactCoordinates;
+    
+    private String artifactPath;
```
Comment (sbespalov) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/domain/ArtifactEntry.java]: @carlspring you may want to note this denoramalization. This is needed to create unique index for ArtifactEntry.  The index in turn is needed in order to quickly determine the presence of the required entity.

Comment (sbespalov) [reply strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/domain/ArtifactEntry.java]: I was not able to find the best way :(

File: strongbox-commons/src/main/java/org/carlspring/strongbox/artifact/coordinates/AbstractArtifactCoordinates.java [hunk 7]
```diff
@@ -5,6 +5,8 @@
 
 import org.carlspring.strongbox.data.domain.GenericEntity;
 
+import edu.emory.mathcs.backport.java.util.Collections;
```
Comment (carlspring) [inline strongbox-commons/src/main/java/org/carlspring/strongbox/artifact/coordinates/AbstractArtifactCoordinates.java]: This is a incorrect import.

**Comment by carlspring** [comment]
Thanks! :)
